### DISCUSSION TITLE
Instrument Remix app code

### DISCUSSION THREAD

**Discussion by nicholaschiang** [remix-run/remix#2284] [Q&A]
I'm writing Cypress integration tests and want to track my tests' code coverage. Previously, I've relied on `babel-plugin-istanbul` and `nyc` to generate code coverage reports (that are then sent, during CI, to a service like CodeCov), but that doesn't seem possible with [Remix](https://remix.run/docs/en/v1/pages/technical-explanation#compiler) as it uses `esbuild` to compile everything.

@mjackson @ryanflorence I've looked over some of your past example apps (e.g. `remix-ecommerce`, `jokes-app`, etc) and while many of them use Cypress for integration testing, none of them (that I could find) track code coverage. Is there an accepted way to do so?

I've also taken a look at the [`esbuild`](https://github.com/evanw/esbuild) issues and found a couple (e.g. [#516](https://github.com/evanw/esbuild/issues/516), [#184](https://github.com/evanw/esbuild/issues/184)) that are relevant but don't provide any solutions (besides asserting that instrumenting code is beyond the scope of `esbuild` and thus isn't planned on being supported anytime soon... which is unfortunate because using Remix seems like a lock-in to `esbuild` and thus a lock-out of code coverage tracking). :sob:

**Comment by miku86** [comment]
@nicholaschiang, tried to make it work for some hours, too, but found the some problems you had.
Any news from your side regarding test coverage?

**Reply by nicholaschiang** [reply]
nope, i think i ended up realizing that in order to get code coverage i'll have to implement my own compiler. feel free to search the remix discord for this (as that's where the discussion ended up happening i think). @ryanflorence had some suggestions there.

**Comment by Visible-Radio** [comment]
In 2024 is there a way to do this? Would like to get coverage out of playwright. Everywhere I look the answer seems to be `babel-plugin-istanbul` and `nyc` - but as mentioned by OP this is not tenable.

**Comment by SylRob** [comment]
Hi, I am also in the same situation, I would like to have a code coverage report (with playwright).

**Comment by cenfun** [comment]
Here is a example to generate code coverage for Remix
https://github.com/cenfun/remix-with-playwright
